Assignment 2 solution

cba33 opened the pull request.

cba33 committed:
Assignment 2 solution

brzozan commented on BinaryTree.java:
```diff
@@ -0,0 +1,137 @@
+ * Created by basescu on 29.05.17.
+ */
+public class BinaryTree {
+    private Integer info;
```
Could be just int, unless you need to handle "value not present" scenarios, which does not seem to be the case. Same for other occurrences of Integer below.

Unboxed types (int, long, boolean etc) are much more efficient than boxed ones (Integer, Long, Boolean etc) and should be preferred unless really needed.

Please also consider making this field (probably as well as left and right) final. They are assigned only in constructor and never modified, so it is usually good to mark them as such so that compiler can detect unintended overwrites.

cba33 replied:
Thank you, good tip, I changed it.

brzozan commented on BinaryTree.java:
```diff
@@ -0,0 +1,137 @@
+    }
+
+    private String printAncestors(Integer key, String ancestors) {
+        if(this.info == key)
```
Please always use {} around code blocks after if, for etc. It is not required by language itself, but it is very easy to make mistake in code that does not use them, which can lead to very serious bugs.

And a stylistic nitpick: Java coding convention is to typically put a space between keyword and opening parentheses. Oracle's style guide does not mention it explicitly, but [Google](https://google.github.io/styleguide/javaguide.html) coding conventions do and I believe most of Java codebase includes them.

cba33 replied:
Thanks, I changed it to that!

brzozan commented on BinaryTree.java:
```diff
@@ -0,0 +1,137 @@
+        return info;
+    }
+
+    public String printAncestors(Integer key) {
```
What if at some point we would like to do something interesting with ancestors rather than just print them?

Could you please refactor code to separate finding ancestors from printing them?

cba33 replied:
Sounds good, I changed it.

brzozan commented on BinaryTree.java:
```diff
@@ -0,0 +1,137 @@
+        if(this.right == null)
+            return leftRes;
+
+        if(this.left.contains(tree1) && this.right.contains(tree2))
```
This causes multiple tree traversals (one in recursive calls to lowestCommonAncestor, one in contains). Could code be changed to avoid that?

Hint: Could results of finding ancestors of a single node somehow be used to do that?

brzozan commented on BinaryTree.java:
```diff
@@ -0,0 +1,137 @@
+        System.out.println(bt.printAncestors(5));
+
+        // Question 2
+        System.out.println(bt.lowestCommonAncestor(new BinaryTree(5, null, null), new BinaryTree(14, null, null)).getInfo());
```
Finding common ancestor has quite a few edge cases that need to be handled. This println only covers one case. Additionally it has to be then read by human and manually checked if printed value is correct, which is error prone.

Could you change the code so that lowestCommonAncestor is called with different values covering different edge cases (like key not present in the tree, one key being ancestor of the other etc) and for each of them prints status, whether returned value was correct or not?
For bonus points you could look at [JUnit](https://github.com/junit-team/junit4/wiki/Getting-started), but writing your own solution is perfectly acceptable.

Note, that testing code (including main() method) should not be inside BinaryTree class, as the only concern of BinaryTree should be implementation of binary tree and methods that operate on it. Testing code should live in a separate class, for example BinaryTreeTest or something similar.

cba33 replied:
Thanks for the tip, I used JUnit for the code version.

brzozan commented on BinaryTree.java:
```diff
@@ -0,0 +1,137 @@
+        return right;
+    }
+
+    private BinaryTree lowestCommonAncestor(BinaryTree tree1, BinaryTree tree2) {
```
Using BinaryTree as inputs to this method is quite inconvenient and potentially confusing - a person might think that they can pass multi-level trees here and that algorithm will match them against the tree contained in this. For example if this contains a tree (1, (2, null, null), (3, (4, null, null), (5, null, null))) one might try to find common ancestor of (3, (2, null, null), null) (note, that such subtree does not exist in our main tree) and (2, null, null), and would be surprised that common ancestor is found even though first subtree does not exist.

Since only info of the root node of each parameter tree is used, why not just use int info1 and int info2 here?

brzozan commented on BinaryTree.java:
```diff
@@ -0,0 +1,137 @@
+        return (containedLeft || containedRight);
+    }
+
+    private boolean isIdentical(BinaryTree tree) {
```
Is checking for identical really needed?

cba33 replied:
I thought we wanted to find the ancestor of two subtrees: for instance because keys are not unique, but they're unique if we provide the entire subtree. I changed it to keys in the new code version, assuming keys are unique.

brzozan reviewed (changes requested):
Thank you! I have added a few comments to the code.

cba33 committed:
Addressed requested changes

- Smaller number of recursive calls in commonAncestor (still n^2 overall)
- Tests with JUnit
- All the other remarks also addressed. Thanks!

brzozan commented on BinaryTreeTest.java:
```diff
@@ -0,0 +1,106 @@
+        );
+    }
+
+    @org.junit.jupiter.api.Test
```
Why not import org.junit.jupiter.api.Test and use `@Test` instead of fully qualified class name?

brzozan commented on BinaryTreeTest.java:
```diff
@@ -0,0 +1,106 @@
+        try {
+            String ancestors = bt.printAncestors(14);
+            assertEquals(ancestors, "9 16");
+        } catch(InexistentKeyException e) {
```
No need to catch exceptions unless they are expected. By default JUnit assumes that test should not throw, so it will treat exception as failed test.
Just declare your test method with 'throws Exception' and let exception happen.

brzozan commented on BinaryTreeTest.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    @org.junit.jupiter.api.Test
+    void findAncestors() {
```
It is a good practice to have only a single assertion in each test method and to name the method in a descriptive way (either describing the test, like 'findAncestors_leafNode', or describing expected behavior, like 'findAncestorsForLeafReturnsWholeBranch'; in this case probably the first schema is better, since behavior is quite trivial).

Could you please split findAncestors into multiple, single-assertion methods?

brzozan commented on BinaryTreeTest.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    @org.junit.jupiter.api.Test
+    void lowestCommonAncestor() {
```
Please apply the same comments to the tests for lowestCommonAncestor.

brzozan commented on BinaryTree.java:
```diff
@@ -0,0 +1,148 @@
+            sb.append(crtNode.getInfo());
+            sb.append(" ");
+        }
+        sb.deleteCharAt(sb.length() - 1);
```
That seems a bit too complex for the purpose, especially that it makes it necessary to specifically check for empty ancestors list.

Could code be somehow changed to add separator only when necessary rather than delete unnecessary one?

brzozan commented on BinaryTree.java:
```diff
@@ -0,0 +1,148 @@
+        }
+
+        /* initialize String Builder with probably enough space - 6 chars per key (including space separator) */
+        StringBuilder sb = new StringBuilder(6 * ancestors.size());
```
Providing initial size is usually not needed - StringBuilder does good job efficiently growing internal buffer as needed, so unless you know for sure that it is a bottleneck (via profiling), it is better to opt simpler code that uses defaults.

brzozan commented on BinaryTree.java:
```diff
@@ -0,0 +1,148 @@
+        return info;
+    }
+
+    public String printAncestors(int key) throws InexistentKeyException{
```
public String printAncestors has quite limited use - it is useful only for people who are interested in receiving string with a fixed representation of ancestors. It would be better to have public API that just returns list of ancestors (preferably "unpacked" from BinaryTree objects) and a helper method (not necessarily in BinaryTree) that converts it to String.

And a nitpick: space before {

brzozan commented on BinaryTree.java:
```diff
@@ -0,0 +1,148 @@
+
+class InexistentKeyException extends Exception {
+    @Override
+    public String getMessage() {
```
Alternatively you could define default constructor that would then call `super("Key does not exist in tree")`. This would be a bit cleaner, since internal state of the exception object would be consistent.

brzozan commented on BinaryTree.java:
```diff
@@ -0,0 +1,148 @@
+        return sb.toString();
+    }
+
+    private LinkedList<BinaryTree> findAncestors(int key) throws InexistentKeyException {
```
It would be a bit better to return `List<BinaryTree>` rather than `LinkedList`. For the user of this method it does not matter which implementation of list is returned and if you decide to change it to `ArrayList` in the future you will need to change only one place rather than all clients that rely on the fact that initially it returned `LinkedList`.

brzozan replied:
Separate efficiency comment: throwing exception is a nice public API design, but exceptions have some overhead, so for internal methods that are called many times it is better to just return `null`, `Optional` or empty list.

Exceptions should be reserved for exceptional situations, while in this internal method you actually do expect it to be thrown at least once per call (key is either in left or in right branch), multiplied by recursion.

brzozan commented on BinaryTree.java:
```diff
@@ -0,0 +1,148 @@
+                leftAncestors.add(this);
+                return leftAncestors;
+            } catch (InexistentKeyException e) {
+
```
Nitpick: please remove empty line.

brzozan commented on BinaryTree.java:
```diff
@@ -0,0 +1,148 @@
+        /* find ancestors of key 1, take each of them and see if the tree rooted there contains key2 */
+        LinkedList<BinaryTree> ancestorsKey1 = findAncestors(key1);
+
+        if(ancestorsKey1.size() == 0) {
```
Formatting nitpick: always space after `if`, `for` etc.

And another nitpick: ancestorsKey1.isEmpty() instead of size check is a bit more readable, as it immediately conveys the purpose of the check.

brzozan commented on BinaryTree.java:
```diff
@@ -0,0 +1,148 @@
+            if (i > 0) {
+                prevAncestor = ancestorsKey1.get(i-1).getInfo();
+            }
+            else {
```
Formatting nitpick: typically `else` goes in the same line as bracket closing `if` part:

```
if (...) {
  ...
} else {
  ...
}
```

brzozan commented on BinaryTree.java:
```diff
@@ -0,0 +1,148 @@
+
+        for (int i = 0 ; i < ancestorsKey1.size() ; i++) {
+
+            BinaryTree intermedNode = ancestorsKey1.get(i);
```
What is the complexity of getting i-th element of a linked list? :)

brzozan commented on BinaryTree.java:
```diff
@@ -0,0 +1,148 @@
+            }
+
+
+            if (intermedNode.left != null) {
```
This logic with multi-level ifs is quite complex. Could it be simplified? What if you had list of ancestors of both of key1 and of key2? Whould that help?

Side note: solution with lists of ancestors is technically not allowed, since it copies parts of the tree to a separate data structure, which was forbidden (in a slightly convoluted way) by assignment specification. I missed this limitation myself and shouldn't have asked you to do it, and I apologize for doing so, but since we've already taken this route, let's get it right :)

In the future please feel free to disagree if I ask you to do something that seems to be against requirements.

brzozan reviewed (changes requested):
Thanks for doing requested changes! It starts to look really good.
We're almost done, just a few more comments.
